test/qdmi/test_driver.cpp

```diff
@@ -0,0 +1,379 @@
+              QDMI_ERROR_INVALIDARGUMENT)
+        << "The MAX property is not a valid value for any device.";
+  }
+}
+
+TEST_P(DriverTest, QueryOperations) {
+  size_t operationsSize = 0;
+  ASSERT_EQ(QDMI_device_query_device_property(device,
+                                              QDMI_DEVICE_PROPERTY_OPERATIONS,
+                                              0, nullptr, &operationsSize),
+            QDMI_SUCCESS)
+      << "Failed to get the size to retrieve the operations.";
+  std::vector<QDMI_Operation> operations(operationsSize /
+                                         sizeof(QDMI_Operation));
+  ASSERT_EQ(QDMI_device_query_device_property(
+                device, QDMI_DEVICE_PROPERTY_OPERATIONS, operationsSize,
+                static_cast<void*>(operations.data()), nullptr),
+            QDMI_SUCCESS)
+      << "Failed to retrieve the operations.";
+  for (auto* const op : operations) {
+    size_t namesSize = 0;
+    ASSERT_EQ(QDMI_device_query_operation_property(
+                  device, op, 0, nullptr, 0, nullptr,
+                  QDMI_OPERATION_PROPERTY_NAME, 0, nullptr, &namesSize),
+              QDMI_SUCCESS)
+        << "Failed to get the length of the operation's name.";
+    std::string name(namesSize - 1, '\0');
+    ASSERT_EQ(
+        QDMI_device_query_operation_property(device, op, 0, nullptr, 0, nullptr,
+                                             QDMI_OPERATION_PROPERTY_NAME,
+                                             namesSize, name.data(), nullptr),
+        QDMI_SUCCESS)
+        << "Failed to retrieve the operation's name.";
+    EXPECT_FALSE(name.empty())
+        << "Device must provide a non-empty name for every operation.";
+
+    size_t numParams = 0;
+    ASSERT_EQ(QDMI_device_query_operation_property(
+                  device, op, 0, nullptr, 0, nullptr,
+                  QDMI_OPERATION_PROPERTY_PARAMETERSNUM, sizeof(size_t),
+                  &numParams, nullptr),
+              QDMI_SUCCESS)
+        << "Failed to query number of parameters for operation.";
+
+    double duration = 0;
+    double fidelity = 0;
+    std::random_device rd;
+    std::mt19937 gen(rd());
+    std::uniform_real_distribution<double> dis(0.0, 1.0);
+    std::vector<double> params(numParams);
+    for (auto& param : params) {
+      param = dis(gen);
+    }
+    EXPECT_EQ(QDMI_device_query_operation_property(
+                  device, op, 0, nullptr, numParams, params.data(),
+                  QDMI_OPERATION_PROPERTY_DURATION, sizeof(double), &duration,
+                  nullptr),
+              QDMI_SUCCESS)
+        << "Failed to query duration for operation " << name << ".";
+    EXPECT_EQ(QDMI_device_query_operation_property(
+                  device, op, 0, nullptr, numParams, params.data(),
+                  QDMI_OPERATION_PROPERTY_FIDELITY, sizeof(double), &fidelity,
+                  nullptr),
+              QDMI_SUCCESS)
+        << "Failed to query fidelity for operation " << name << ".";
```

Comment:
Also here, I think we cannot enforce these to be defined (IQM devices for example won't define a duration for the foreseeable future). But if they are defined, they should have sensible values. Durations should be greater equal zero and fidelities should be between 0. And 1.